src/Models/class-smart-link.php

```diff
@@ -94,28 +118,277 @@ class Smart_Link extends Base_Model {
 	 * @param string $title The title of the suggested link.
 	 * @param string $text The text of the suggested link.
 	 * @param int    $offset The offset/position for the suggested link.
+	 * @param int    $post_id The post ID of the suggested link.
+	 */
+	public function __construct(
+		string $href,
+		string $title,
+		string $text,
+		int $offset,
+		int $post_id = 0
+	) {
+		$this->set_href( $href );
+		$this->title          = $title;
+		$this->text           = $text;
+		$this->offset         = $offset;
+		$this->source_post_id = $post_id;
+
+		parent::__construct();
+	}
+
+	/**
+	 * Gets the smart link post object by UID.
+	 *
+	 * @since 3.16.0
+	 *
+	 * @param string $uid The UID of the smart link.
+	 * @return int The ID of the smart link post object.
 	 */
-	public function __construct( string $href, string $title, string $text, int $offset ) {
-		$this->href    = $href;
-		$this->title   = $title;
-		$this->text    = $text;
-		$this->offset  = $offset;
-		$this->applied = false;
+	private function get_smart_link_object_by_uid( string $uid ): int {
+		$cached = wp_cache_get( $uid . $this->source_post_id, 'wp_parsely_smart_link_id' );
+		if ( is_int( $cached ) && 0 !== $cached ) {
+			return $cached;
+		}
 
-		$this->post_id = $this->get_post_id_by_url( $href );
+		$smart_links = new \WP_Query(
+			array(
+				'post_type'      => 'parsely_smart_link',
+				'fields'         => 'ids', // Only get the post IDs to improve performance.
+				'posts_per_page' => 1,
+				'title'          => $uid,
+				// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_tax_query
+				'tax_query'      => array(
+					array(
+						'taxonomy'         => 'smart_link_source',
+						'include_children' => false, // Performance optimization.
+						'field'            => 'name',
+						'terms'            => (string) $this->source_post_id,
+					),
+				),
+			)
+		);
+
+		if ( $smart_links->have_posts() && is_int( $smart_links->posts[0] ) ) {
+			wp_cache_set(
+				$uid . $this->source_post_id,
+				$smart_links->posts[0],
+				'wp_parsely_smart_link_id'
+			);
+			return $smart_links->posts[0];
+		}
 
-		if ( false === $this->post_id ) {
-			$this->post_type = 'external';
+		return 0;
+	}
+
+	/**
+	 * Loads the smart link post object.
+	 *
+	 * @since 3.16.0
+	 *
+	 * @return bool True if the smart link was loaded successfully, false otherwise.
+	 */
+	private function load(): bool {
+		if ( 0 === $this->smart_link_id ) {
+			// Try to get the smart link id from the UID.
+			$this->smart_link_id = $this->get_smart_link_object_by_uid( $this->uid );
+			if ( 0 === $this->smart_link_id ) {
+				$this->exists = false;
+				return false;
+			}
+		}
+
+		$smart_link = get_post( $this->smart_link_id );
+
+		if ( null === $smart_link ) {
+			$this->exists = false;
+			return false;
+		}
+
+		$this->exists  = true;
+		$this->applied = true;
+
+		$this->uid = $smart_link->post_title;
```

Comment:
It is, the UID is stored as the post title in the Smart Link CPT, hence setting here the UID with the post title. This is just to make sure that if the `load()` method loads the smart link using the CPT object id, it would also update the `uid` to the one stored in the database (as a title).